File: 030. Spiral Traverse/SpiralTraverseRecursive.java
import java.util.*;

class Program {

    // O(n) time | O(n) space - where n is the total number of elements
    public static List<Integer> spiralTraverse(int[][] array) {
        if (array.length == 0) {
            return new ArrayList<Integer>();
        }

        var result = new ArrayList<Integer>();
        spiralFill(array, 0, array.length - 1, 0, array[0].length - 1);
        return result;
    }

    public static void spiralFill(int[][] array, int startRow, int endRow, int startCol, int endCol,
            ArrayList<Integer> result) {
        if (startRow > endRow || startCol > endCol) {
            return;
        }

        for (int col = startCol; col <= endCol; col++) {
            result.add(array[startRow][col]);
        }
        for (int row = startRow + 1; row <= endRow; row++) {
            result.add(array[row][endRow]);
        }
        for (int col = endCol - 1; col >= startCol; col--) {
            if (startRow == endRow)
                break;
            result.add(array[endRow][col]);
        }
        for (int row = endRow - 1; row >= startRow + 1; row--) {
            if (startCol == endCol)
                break;
            result.add(array[row][startCol]);
        }
        spiralFill(array, startRow + 1, endRow - 1, startCol + 1, endCol - 1, result);
    }
}

File: 030. Spiral Traverse/SpiralTraverseRecursive.kt
package com.algoexpert.program

// O(n) time | O(n) space - where n is the total number of elements in the array
fun spiralTraverse(array: List<List<Int>>): List<Int> {
    if (array.size == 0) return mutableListOf<Int>()

    var result = mutableListOf<Int>()
    spiralFill(array, 0, array.size - 1, 0, array[0].size - 1, result)
    return result
}

fun spiralFill(
        array: List<List<Int>>,
        startRow: Int,
        endRow: Int,
        startCol: Int,
        endCol: Int,
        result: MutableList<Int>
) {
    if (startRow > endRow || startCol > endCol) return

    for (col in startCol..endCol) result.add(array[startRow][col])
    for (row in startRow + 1..endRow) result.add(array[row][endCol])
    for (col in endCol - 1 downTo startCol) {
        if (startRow == endRow) break
        result.add(array[endRow][col])
    }
    for (row in endRow - 1 downTo startRow + 1) {
        if (startCol == endCol) break
        result.add(array[row][startCol])
    }
    spiralFill(array, startRow + 1, endRow - 1, startCol + 1, endCol - 1, result)
}
